src/snarks/groth16/mpc/mpc_utils.tcc

```diff
@@ -0,0 +1,257 @@
+#ifndef __ZETH_SNARKS_GROTH16_MPC_UTILS_TCC__
+#define __ZETH_SNARKS_GROTH16_MPC_UTILS_TCC__
+
+#include "evaluator_from_lagrange.hpp"
+#include "mpc_utils.hpp"
+#include "multi_exp.hpp"
+#include "phase2.hpp"
+#include "util.hpp"
+
+#include <algorithm>
+#include <exception>
+#include <libfqfft/evaluation_domain/domains/basic_radix2_domain_aux.tcc>
+
+namespace libzeth
+{
+
+template<typename ppT>
+srs_mpc_layer_L1<ppT>::srs_mpc_layer_L1(
+    libff::G1_vector<ppT> &&T_tau_powers_g1,
+    libff::G1_vector<ppT> &&A_g1,
+    libff::G1_vector<ppT> &&B_g1,
+    libff::G2_vector<ppT> &&B_g2,
+    libff::G1_vector<ppT> &&ABC_g1)
+    : T_tau_powers_g1(std::move(T_tau_powers_g1))
+    , A_g1(std::move(A_g1))
+    , B_g1(std::move(B_g1))
+    , B_g2(std::move(B_g2))
+    , ABC_g1(std::move(ABC_g1))
+{
+}
+
+template<typename ppT> size_t srs_mpc_layer_L1<ppT>::degree() const
+{
+    return T_tau_powers_g1.size() + 1;
+}
+
+template<typename ppT> bool srs_mpc_layer_L1<ppT>::is_well_formed() const
+{
+    return libzeth::container_is_well_formed(T_tau_powers_g1) &&
+           libzeth::container_is_well_formed(A_g1) &&
+           libzeth::container_is_well_formed(B_g1) &&
+           libzeth::container_is_well_formed(B_g2) &&
+           libzeth::container_is_well_formed(ABC_g1);
+}
+
+template<typename ppT>
+void srs_mpc_layer_L1<ppT>::write(std::ostream &out) const
+{
+    using G1 = libff::G1<ppT>;
+    using G2 = libff::G2<ppT>;
+    check_well_formed(*this, "mpc_layer1 (write)");
```

Comment:
In `util.tcc` there are 2 functions:
```cpp
template<typename T> void check_well_formed(const T &v, const char *name)
{
    if (!v.is_well_formed()) {
        throw std::invalid_argument(std::string(name) + " not well-formed");
    }
}

template<typename T> void check_well_formed_(const T &v, const char *name)
{
    if (!is_well_formed(v)) {
        throw std::invalid_argument(std::string(name) + " not well-formed");
    }
}
```

Why having two very similar functions like that?
What about having a template `StructuredT` to represent structured types for which it makes sense to define a `check_well_formed` method? We can document the functions that are expected to be implemented by a `StructuredT` type to have a sort of interface. Otherwise, it may be worth using abstract classes for that, and use inheritance for all the objects that need to implement a `check_well_formed` method. Doing so makes the code base clearer (seeing a class inheriting from `structured` being an abstract class makes the code base easier to read and clearer as the interfaces are well defined and easier to maintain IMO)

Reply:
For all types we define ourselves (where it makes sense) we include the `is_well_formed` method, so `check_well_formed` can be used.
`check_well_formed_` is for the case where we want to perform this kind of check on a type which is not part of zeth (I think mainly the proving and verification keys in libsnark), where we cannot easily add a new method.

(I am not sure if this has answered your comment here - let me know if I've misunderstood the point.)

Reply:
I see, thanks for the clarification. Maybe we can add a comment to reflect your explanation above into the code?

> (I am not sure if this has answered your comment here - let me know if I've misunderstood the point.)

Yes you answered my first question, thanks.

I still think that using inherited abstract classes can be useful to define our interfaces. What do you think?

Reply:
(To capture the results of a discussion ...) I agree that the abstract base classes approach will better capture these constraints.  We agreed to think about it.

For now, the point about naming seems a good idea and minimal impact.  I've renamed this (and type params in other places) to be more meaningful than just `T`.

Reply:
Thanks, please add comments where relevant to clarify what "interfaces" each type implement (eg: https://github.com/clearmatics/zeth/blob/develop/src/circuits/sha256/sha256_ethereum.hpp#L8)

Reply:
Thanks.  I've added comments describing the new `StructuredT` and `ReadableT` "interfaces" and a comment on each strcture that implements them.